File: go/104-simple-linked-list/simple_linked_list.go
package linkedlist

import (
	"errors"
	"fmt"
)

type Node struct {
	Value int
	next  *Node
	prev  *Node
}

type List struct {
	Head *Node
	Tail *Node
}

func New(elements []int) *List {
	list := &List{}
	for _, e := range elements {
		list.Push(e)
	}
	return list
}

func (l *List) Size() int {
	count := 0
	current := l.Head
	for current != nil {
		count++
		current = current.next
	}
	return count
}

func (l *List) Push(element int) {
	newNode := &Node{Value: element}

	if l.Head == nil {
		l.Head = newNode
		l.Tail = newNode
		return
	}

	newNode.prev = l.Tail
	l.Tail.next = newNode
	l.Tail = newNode
}

func (l *List) Pop() (int, error) {

	if l.Tail == nil {
		return 0, errors.New("empty list")
	}

	val := l.Tail.Value
	if l.Tail.prev == nil {
		l.Head = nil
		l.Tail = nil
		return val, nil
	}

	l.Tail = l.Tail.prev
	l.Tail.next = nil

	return val, nil
}

func (l *List) Array() []int {
	var result []int
	fmt.Println(l.Head)
	current := l.Head
	for current != nil {
		result = append(result, current.Value)
		current = current.next
	}

	return result
}

func (l *List) Reverse() *List {
	reversedList := &List{}
	current := l.Tail
	for current != nil {
		reversedList.Push(current.Value)
		current = current.prev
	}

	return reversedList

}

File: ruby/085-simple-linked-list/simple_linked_list.rb
class Element
  attr_accessor :next
  attr_reader :value

  # @param[Numeric]value
  def initialize(value)
    @value = value
    @next = nil
  end

  def datum
    @value
  end
end

class SimpleLinkedList
  attr_reader :head

  def initialize(arr = [])
    @head = nil
    arr.each { |value| internal_push_value(value) }
  end

  # @param[Element] element
  def push(element)
    element.next = @head
    @head = element
    self
  end

  # @return[Element, nil]
  def pop
    return nil unless @head

    popped_element = @head
    @head = @head.next
    popped_element.next = nil
    popped_element
  end

  # @return[Array]
  def to_a
    result = []
    current = @head
    while current
      result << current.datum
      current = current.next
    end
    result
  end

  # @return[SimpleLinkedList]
  def reverse!
    previous = nil
    current = @head
    while current
      next_node = current.next  # Store next node
      current.next = previous   # Reverse current node's pointer
      previous = current        # Move previous one step forward
      current = next_node       # Move current one step forward
    end
    @head = previous            # The last 'previous' is the new head
    self
  end

  private

  def internal_push_value(value)
    element = Element.new(value)
    push(element)
  end
end
